text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way
+to generalise integers *for computers*, but generalising based upon bounds is
+a very natural way for humans to do it, *and* more general. For example, instead
+of `uint<N>` and `int<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a
+percentage could be represented exactly as `int<0..=100>`. Whether an integer is
+signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot harder to
+implement, and could be added in the future as an extension. Longer-term, we
+could for example guarantee that `int<0..=2.pow(N)-1>` is equivalent to
+`uint<N>`, and `int<-2.pow(N)..=2.pow(N)-1>` is equivalent to `int<N>`, ensuring
+that ultimately, we could replace literals with our bounded types.
+
+Again, how these bounded types work would have to be fleshed out in a future
+RFC.
+
+## Integer traits
+
+Previously, Rust had traits which generalised the integer types and their
+methods, but these were ultimately removed in favour of inherent methods.
+Going with a generic `uint<N>` over an `Int` trait would avoid the problem of
+determining which methods are suited for these traits; instead, the `uint<N>`
+type would have all of them.
+
+Additionally, having these traits does not allow non-power-of-two `uint<N>`
+outright, and so, this feature is a strict superset of that.
+
+Additionally, having separate `uint<N>` and `int<N>` types also solves the
+problem of generalising signed and unsigned integers: everything would require
+one impl for signed integers, and one for unsigned. This would ensure that these
+types have exactly the same behaviour for all `N`, only differing in the upper
+bound on the number of bits.
+
+## Macro-based bitfield structs
+
+Many proposals for bitfields involved using macros to create structs, which
+would ultimately allow the underlying mechanisms for bitfield-based structs to
+be decided by the ecosystem, rather than the language itself.
+
+This is actually orthogonal to the proposal, as having language-level `uint<N>`
+types means that such macros would be able to return exactly the type necessary
+for each field. While something like `repr(bitpacked)` would be very far off in
+terms of stabilisation, these macros could simply get and set `uint<N>` values,
+and users could determine how to work with them without having to worry about
+checking for overflow or manually implementing code which has already been
+written.
+
+So, as far as this RFC is concerned, this is again orthogonal to the existing
+RFC.
+
+## Offering as a library
+
+Once const generics and specialisation are implemented and stable, almost all of
+this could be offered as a crate which offers `uint<N>` and `int<N>` types. I
+won't elaborate much on this because I feel that there are many other
```

Comment:
You should :) Do we know that it is actually implementable as a library?

Reply:
I will try to get to this this weekend.